providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,734 @@
+                if (!ossl_assert(bytes_written 
+                                 < ds - ctx->info->alg[i].pubkey_length_bytes))
+                    return 0;
+                q += ctx->info->alg[i].pubkey_length_bytes;
+                bytes_written += ctx->info->alg[i].pubkey_length_bytes;
+                OSSL_PARAM_set_all_unmodified(prms);
+                if (!EVP_PKEY_CTX_get_params(ctx->ctxs[i], prms)
+                        || !OSSL_PARAM_modified(prms))
+                    return 0;
+            }
+            break;
+
+        default:
+            if (!ctx_subalg_get_param(ctx->info->num_algs, ctx->ctxs, p))
+                return 0;
+            break;
+        }
+    }
+    return 1;
+}
+
+int ossl_hybrid_set_ctx_params(HYBRID_PKEY_CTX *ctx, const OSSL_PARAM params[])
+{
+    char *str = NULL;
+    const OSSL_PARAM *p;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    if (!ctx_subalg_set_param(ctx->info->num_algs, ctx->ctxs, params))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL)
+        if (strcmp(p->key, OSSL_PKEY_PARAM_PROPERTIES) == 0) {
+            if (!OSSL_PARAM_get_utf8_string(p, &str, SIZE_MAX))
+                return 0;
+            OPENSSL_free(ctx->propq);
+            ctx->propq = str;
+        }
+    return 1;
+}
+
+int ossl_hybrid_kmgmt_gen_get_params(void *vctx, OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_get_ctx_params(ctx, params);
+}
+
+int ossl_hybrid_kmgmt_gen_set_params(void *vctx, const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+void *ossl_hybrid_kmgmt_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY *key = OPENSSL_zalloc(sizeof(*key));
+    unsigned int i;
+
+    if (key == NULL)
+        return NULL;
+
+    key->libctx = ctx->libctx;
+    key->info = ctx->info;
+    if (ctx->propq != NULL) {
+        key->propq = OPENSSL_strdup(ctx->propq);
+        if (key->propq == NULL)
+            goto err;
+    }
+    INIT_ACCUMULATE_HYBRID_NUMBERS(key);
+    for (i = 0; i < ctx->info->num_algs; i++) {
+        if (EVP_PKEY_keygen(ctx->ctxs[i], key->keys + i) <= 0)
+            goto err;
+        ACCUMULATE_HYBRID_NUMBERS(key, i);
+    }
+    return key;
+ err:
+    ossl_hybrid_kmgmt_free(key);
+    return NULL;
+}
+
+void ossl_hybrid_kmgmt_gen_cleanup(void *vctx)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    unsigned int i;
+
+    if (ctx != NULL) {
+        for (i = 0; i < ctx->info->num_algs; i++)
+            EVP_PKEY_CTX_free(ctx->ctxs[i]);
+        OPENSSL_free(ctx->propq);
+        OPENSSL_free(ctx);
+    }
+}
+
+int ossl_hybrid_kmgmt_import(void *keydata, int selection,
+                             const OSSL_PARAM params[])
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)keydata;
```

Comment:
This format seems to imply that exported data is not persisted.. It seems any key string added would break the import?

Reply:
Yep.  This is a purely transient format so that the OSSL_PARAMs from different algorithms can be handled sanely.

At some point in the future, we might end up with encoders and decoders which would deal with non-transient keying.